src/vsr/journal.zig

```diff
@@ -487,39 +584,56 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
 
             // Do not use this pointer beyond this function's scope, as the
             // header memory may then change:
-            const exact = self.entry_for_op_exact_with_checksum(op, checksum) orelse {
+            const exact = self.header_with_op_and_checksum(op, checksum) orelse {
                 self.read_prepare_log(op, checksum, "no entry exactly");
                 callback(replica, null, null);
                 return;
             };
 
-            if (self.faulty.bit(op)) {
-                assert(self.dirty.bit(op));
+            const slot = self.slot_with_op_and_checksum(op, checksum).?;
+            if (self.faulty.bit(slot)) {
+                assert(self.dirty.bit(slot));
 
                 self.read_prepare_log(op, checksum, "faulty");
                 callback(replica, null, null);
                 return;
             }
 
-            if (self.dirty.bit(op)) {
+            if (self.dirty.bit(slot)) {
                 self.read_prepare_log(op, checksum, "dirty");
                 callback(replica, null, null);
                 return;
             }
 
-            const physical_size = vsr.sector_ceil(exact.size);
-            assert(physical_size >= exact.size);
-
-            const message = replica.message_bus.get_message();
-            defer replica.message_bus.unref(message);
-
             // Skip the disk read if the header is all we need:
             if (exact.size == @sizeOf(Header)) {
+                const message = replica.message_bus.get_message();
+                defer replica.message_bus.unref(message);
+
                 message.header.* = exact.*;
                 callback(replica, message, destination_replica);
                 return;
             }
 
+            self.read_prepare_with_op_and_checksum(callback, op, checksum, destination_replica);
+        }
+
+        /// Read a prepare from disk, but there is no in-memory header.
+        pub fn read_prepare_with_op_and_checksum(
+            self: *Self,
+            callback: fn (replica: *Replica, prepare: ?*Message, destination_replica: ?u8) void,
+            op: u64,
+            checksum: u128,
+            destination_replica: ?u8,
+        ) void {
+            const replica = @fieldParentPtr(Replica, "journal", self);
+            const slot = self.slot_for_op(op);
+            assert(checksum == self.prepare_checksums[slot.index]);
+            assert(checksum != 0);
```

Comment:
Implemented in https://github.com/coilhq/tigerbeetle/pull/91/commits/a8a086f0674fd3e0756b0698b5032475ed4796cb. Due to a Zig bug we need to use `[]u128` (paired with a `[]bool` instead of `[]?u128` anyway.